railties/test/test_unit/reporter_test.rb

```diff
@@ -152,7 +152,7 @@ def woot; end
       @reporter = Rails::TestUnitReporter.new @output, color: true, output_inline: true
       record(failed_test)
 
-      expected = %r{\e\[31mF\e\[0m\n\n\e\[31mFailure:\nTestUnitReporterTest::ExampleTest#woot \[test/test_unit/reporter_test.rb:\d+\]:\nboo\n\e\[0m\n\nbin/rails test .*test/test_unit/reporter_test.rb:\d+\n\n}
+      expected = %r{\e\[31mF\e\[0m\n\n\e\[31mFailure:\nTestUnitReporterTest::ExampleTest#woot \[test_unit/reporter_test.rb:\d+\]:\nboo\n\e\[0m\n\n#{test_run_command_regex} .*test/test_unit/reporter_test.rb:\d+\n\n}
```

Comment:
@byroot The failure is on this change. 
It looks like the path displayed by the error is different on CI than when running locally with `bin/test`. 
Does it mean the logic is broken, or should I make the `test/` prefix optional in the regular expression?

```suggestion
      expected = %r{\e\[31mF\e\[0m\n\n\e\[31mFailure:\nTestUnitReporterTest::ExampleTest#woot \[(test/)?test_unit/reporter_test.rb:\d+\]:\nboo\n\e\[0m\n\n#{test_run_command_regex} .*test/test_unit/reporter_test.rb:\d+\n\n}
```

Reply:
Update: I figured (some of) it out.

When I run the test with `bin/test` locally:

- When running this test, `Rails.root` is `/path/to/repo/rails/railties/test`. (~~I can't say I understand why, yet.~~)
  This is due to `test/abstract_unit.rb` defining the Rails application (`TestApp::Application`) to have its root there:
    https://github.com/rails/rails/blob/ff42150d2f2c8df2b9fe21d97fea23fef3810949/railties/test/abstract_unit.rb#L17-L21
- `Rails::BacktraceCleaner` will strip away `Rails.root` from backtrace locations:
  https://github.com/rails/rails/blob/ff42150d2f2c8df2b9fe21d97fea23fef3810949/railties/lib/rails/backtrace_cleaner.rb#L16-L17
- Minitest applies the backtrace cleaner when converting a `Minitest::Result` to a string, which we do when passing `result` to `#color_output` [here](https://github.com/rails/rails/blob/ff42150d2f2c8df2b9fe21d97fea23fef3810949/railties/lib/rails/test_unit/reporter.rb#L30).
- The output ends up being `Failure:\nTestUnitReporterTest::ExampleTest#woot [test_unit/reporter_test.rb:153]` (the path is not prefixed with `test/`.)

Now what I have yet to understand is why, when running on CI, would `Rails.root` be different, or would different backtrace cleaners be applied?

Reply:
Alright, I can reproduce the error locally by running tests the way CI does, instead of `bin/test`:
```sh-session
ruby -w -Itest -Ilib -I../activesupport/lib -I../actionpack/lib -I../actionview/lib -I../activemodel/lib test/test_unit/reporter_test.rb
```

When running the test like this, the path in the output is prefixed with `test/`. `Rails.root` is the same though, so maybe I should look at the backtrace filter more closely.

Reply:
Looks like the problem is located even before backtrace filters have a say?

```rb
# When running bin/test test/test_unit/reporter_test.rb:150 
result.failure.backtrace.first
#=> "/private/tmp/rails/railties/test/test_unit/reporter_test.rb:184:in `failed_test'"

# When running ruby -w -Itest -Ilib -I../activesupport/lib -I../actionpack/lib -I../actionview/lib -I../activemodel/lib test/test_unit/reporter_test.rb
result.failure.backtrace.first
#=> "test/test_unit/reporter_test.rb:184:in `failed_test'"
```

Running `bin/test`, I get backtrace frames with absolute paths, whereas I get relative paths when running with the plain `ruby` command.

The relative paths do not get cleaned, whereas the absolute ones get cleaned based on `Rails.root` being `rails/railties/test`.